lib/QoreSerializable.cpp

```diff
@@ -0,0 +1,1711 @@
+
+        case QSSDT_LIST:
+            return deserializeListFromStream(reader, xsink);
+
+        case QSSDT_FLOAT:
+            return deserializeFloatFromStream(reader, xsink);
+
+        case QSSDT_NUMBER:
+            return deserializeNumberFromStream(reader, xsink);
+
+        case QSSDT_ABSDATE:
+            return deserializeAbsDateFromStream(reader, xsink);
+
+        case QSSDT_RELDATE:
+            return deserializeRelDateFromStream(reader, xsink);
+
+        case QSSDT_BINARY:
+            return deserializeBinaryFromStream(reader, xsink);
+
+        case QSSDT_NULL:
+            return &Null;
+
+        case QSSDT_NOTHING:
+            return QoreValue();
+
+        default:
+            break;
+    }
+
+    xsink->raiseException("DESERIALIZATION-ERROR", "invalid serialization type code " QLLD, code);
+    return QoreValue();
+}
+
+int64 QoreSerializable::readIntFromStream(ExceptionSink* xsink, StreamReader& reader, const char* type, bool can_be_negative) {
+    // read int code
+    int64 code = reader.readi1(xsink);
+    if (*xsink) {
+        return 0;
+    }
+
+    if (!CODE_IS_INT(code)) {
+        xsink->raiseException("DESERIALIZATION-ERROR", "expecting integer code for the %s value; got invalid code %d instead", type, code);
+        return 0;
+    }
+
+    int64 size = deserializeIntFromStream(reader, code, xsink);
+    if (*xsink) {
+        return 0;
+    }
+
+    if (!can_be_negative && size < 0) {
+        xsink->raiseException("DESERIALIZATION-ERROR", "stream data gives an invalid %s value (" QLLD ")", size);
+        return 0;
+    }
+
+    return size;
+}
+
+QoreHashNode* QoreSerializable::deserializeHashFromStream(StreamReader& reader, int64 code, ExceptionSink* xsink) {
+    const TypedHashDecl* thd;
+    int64 size;
+    if (code == QSSDT_HASHDECL) {
+        // read hashdecl path string
+        QoreString path_str;
+        if (readStringFromStream(reader, path_str, "hashdecl tag", xsink)) {
+            return nullptr;
+        }
+
+        const QoreNamespace* pns = nullptr;
+        thd = getProgram()->findHashDecl(path_str.c_str(), pns);
+        if (!thd) {
+            xsink->raiseException("DESERIALIZATION-ERROR", "stream data indicates that a '%s' typed hash should be deserialized, but no such typed hash (hashdecl) could be found in the current Program object", path_str.c_str());
+            return nullptr;
+        }
+    }
+    else {
+        assert(code == QSSDT_HASH);
+        thd = nullptr;
+    }
+
+    // now read in the number of hash keys
+    size = readIntFromStream(xsink, reader, "hash size");
+    if (*xsink) {
+        return nullptr;
+    }
+
+    ReferenceHolder<QoreHashNode> h(new QoreHashNode, xsink);
+
+    // read in hash keys
+    for (int64 i = 0; i < size; ++i) {
+        // read key type (must be string)
+        code = reader.readi1(xsink);
+        if (*xsink) {
+            return nullptr;
+        }
+        if (code != QSSDT_UTF8_STRING && code != QSSDT_STRING) {
```

Comment:
CODE_IS_STRING macro can be used here.

Reply:
I would vote for as few macros as possible, ideally none :)